### ISSUE TITLE
Templated/variable expansion in files

### ISSUE THREAD

**Issue by alexwhitman** [skx/marionette#59]
I have a use case where it would be useful to be able to expand variables within files copied via `source` (and I suppose `source_url` as well). This would probably be disabled by default for backwards compatibility but something like
```
file {
    "source" => "/path/to/file",
    "templated" => true
}
```
could be the way to handle it.

While variables are expanded in `content` statements, expansion in files would be useful for larger (> 1 line) content.

I was going to look at doing this myself but I noticed the environment isn't passed to modules. I'm not sure if, design wise, this would be appropriate?

**Comment by alexwhitman** [comment]
Having thought about it for a bit, a cleaner solution might be to add a `template` attribute to `file` rules, which simply takes the path to a template.
```
file {
    template => "/path/to/template",
    target => "/path/to/target"
}
```

**Comment by skx** [comment]
The approach of using a new type `template` is a good one, and I'd have suggested that too :)

Now you've pointed it out it seems obvious there might be a good reason for modules to have access to the environment.  I'd certainly be OK with changing this:

```
Execute(map[string]interface{}) (bool, error)
```

To include the environment as the first parameter, like so:


```
Execute(env *environment.Environment, args map[string]interface{}) (bool, error)
```
Updating all the existing modules to add that parameter won't take long and it could be useful for other things in the future.

**Comment by skx** [comment]
Closed in #61.
